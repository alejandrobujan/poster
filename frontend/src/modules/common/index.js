export {default as Pager} from "./components/Pager";
export {default as Errors} from "./components/Errors";
export {default as BackLink} from "./components/BackLink";
export {default as UserCard} from "./components/UserCard";
export {default as PostLink} from "./components/PostLink";
export {default as ImagesCard} from "./components/ImagesCard";