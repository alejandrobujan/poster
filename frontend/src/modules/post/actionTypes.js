export const CLEAR_POST_SEARCH = "poster/post/clearPostSearch";
export const FIND_POSTS_COMPLETED = "poster/post/findPostsCompleted";