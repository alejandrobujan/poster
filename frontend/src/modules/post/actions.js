import backend from '../../backend';
import * as actionTypes from './actionTypes';
import * as selectors from './selectors';

const findAllCategoriesCompleted = categories => ({
	type: actionTypes.FIND_ALL_CATEGORIES_COMPLETED,
	categories
});

export const findAllCategories = () => (dispatch, getState) => {

	const categories = selectors.getCategories(getState());

	if (!categories) {

		backend.postService.findAllCategories(
			categories => dispatch(findAllCategoriesCompleted(categories))
		);

	}

}

export const createPost = (title, description, url, price, categoryId, images, type, properties,
	onSuccess, onErrors) => dispatch =>
		backend.postService.createPost(title, description, url, price, categoryId, images, type, properties, onSuccess, onErrors);

const findPostsCompleted = postSearch => ({
	type: actionTypes.FIND_POSTS_COMPLETED,
	postSearch
});


const clearPostSearch = () => ({
	type: actionTypes.CLEAR_POST_SEARCH
});


export const findPosts = criteria => dispatch => {

	dispatch(clearPostSearch());
	backend.postService.findPosts(criteria, result =>
		dispatch(findPostsCompleted({ criteria, result }))
	);

};

export const previousFindPostsResultPage = criteria =>
	findPosts({ page: criteria.page - 1 });

export const nextFindPostsResultPage = criteria =>
	findPosts({ page: criteria.page + 1 });

export const findPostById = id => dispatch => {
	backend.postService.findPostById(id,
		post => dispatch(findPostByIdCompleted(post)));
}

const findPostByIdCompleted = post => ({
	type: actionTypes.FIND_POST_BY_ID_COMPLETED,
	post
});

export const clearPost = () => ({
	type: actionTypes.CLEAR_POST
}); 	

export const deletePost = (id, onSuccess, onErrors) => dispatch => {
	backend.postService.deletePost(id, onSuccess, onErrors);
};


export const updatePostCompleted = post => ({
    type: actionTypes.UPDATE_POST_COMPLETED,
    post
})
    
export const updatePost = (post, onSuccess, onErrors) => dispatch =>
	backend.postService.updatePost(post,
		post => {
			dispatch(updatePostCompleted(post));
			onSuccess();
		}, 
		onErrors);

const markPostAsExpiredCompleted = ({expired}) => ({
	type: actionTypes.MARK_POST_AS_EXPIRED_COMPLETED,
	expired
});

export const markPostAsExpired = (id, expired, onErrors) => dispatch => {
	backend.postService.maskPostAsExpired(id, expired, e => dispatch(markPostAsExpiredCompleted(e)), onErrors);
};	

const ratePostCompleted = ratingCount => ({
	type: actionTypes.RATE_POST_COMPLETED,
	ratingCount
});

export const ratePostPositive = (id, onErrors) => dispatch => {
	backend.ratingService.ratePostPositive(id, ratingCount => dispatch(ratePostCompleted(ratingCount)), onErrors);
};


export const ratePostNegative = (id, onErrors) => dispatch => {
	backend.ratingService.ratePostNegative(id, ratingCount => dispatch(ratePostCompleted(ratingCount)), onErrors);
};
