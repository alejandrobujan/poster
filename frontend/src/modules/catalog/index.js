import * as actions from './actions';
import * as actionTypes from './actionTypes';
import reducer from './reducer';
import * as selectors from './selectors';

export {default as SearchBar} from "./components/SearchBar";
export {default as CategorySelector} from './components/CategorySelector';
export {default as Post} from './components/Post';
export {default as Posts} from './components/Posts';
export {default as Feed} from './components/Feed';

// eslint-disable-next-line
export default {actions, actionTypes, reducer, selectors};


