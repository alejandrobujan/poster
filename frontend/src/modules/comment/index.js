import * as actions from './actions';
import reducer from './reducer';
import * as selectors from './selectors';

export {default as Comments} from './components/Comments';
export {default as Comment} from './components/Comment';

// eslint-disable-next-line
export default  {actions, reducer, selectors};