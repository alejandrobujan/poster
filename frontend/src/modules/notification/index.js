import * as actions from './actions';
import * as actionTypes from './actionTypes';
import reducer from './reducer';
import * as selectors from './selectors';

export {default as Notifications} from "./components/Notifications";
export {default as Notification} from "./components/Notification";

// eslint-disable-next-line
export default {actions, actionTypes, reducer, selectors};


