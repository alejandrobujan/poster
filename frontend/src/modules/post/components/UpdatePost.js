import {useEffect} from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { useParams } from 'react-router-dom';

import * as selectors from '../selectors';
import * as actions from '../actions';

import UpdatePostForm from './UpdatePostForm';

const UpdatePost = ({ min }) => {
	const post = useSelector(selectors.getPost);
	const { id } = useParams();
    const dispatch = useDispatch();
    useEffect(() => {

		const postId = Number(id);

		if (!Number.isNaN(postId)) {
			dispatch(actions.findPostById(postId));
		}

		return () => dispatch(actions.clearPost());

	}, [id, dispatch]);

	if (!post) {
		return null;
	}
    
    return(
		<div>
			<UpdatePostForm min={min}/>
		</div>
	);
};

export default UpdatePost;