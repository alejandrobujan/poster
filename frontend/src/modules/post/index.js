import * as actions from './actions';
import reducer from './reducer';
import * as selectors from './selectors';

export {default as CreatePost} from './components/CreatePost';
export {default as CategorySelector} from './components/CategorySelector';
export {default as Post} from './components/Post';
export {default as Posts} from './components/Posts';
export {default as Feed} from './components/Feed';
export {default as PostDetails} from './components/PostDetails';


// eslint-disable-next-line
export default  {actions, reducer, selectors};