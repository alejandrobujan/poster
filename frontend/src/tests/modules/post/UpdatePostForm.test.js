import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import React from "react";
import { MemoryRouter } from "react-router-dom";
import { Provider } from 'react-redux';

import configureMockStore from 'redux-mock-store'
import thunk from 'redux-thunk'

import renderer from "react-test-renderer";
import { updatePostMock } from "../../state/UpdatePost.mock";
import { updatePostSecondConditionMock } from "../../state/UpdatePostSecondCondition.mock";

import { UpdatePost } from "../../../modules/post";

describe("UpdatePostFormSecondCondition", () => {
	jest.mock('react-router-dom', () => ({
		...jest.requireActual('react-router-dom'), // use actual for all non-hook parts
		useParams: () => ({ id: '1' }),
	  }));

	  const middlewares = [thunk];
	  const mockStore = configureMockStore(middlewares);
	  let store;
	  store = mockStore(updatePostSecondConditionMock);  

	  it("renders correctly", () => {
		const tree = renderer
			.create(
				<Provider store={store}>
					<MemoryRouter>
						<UpdatePost min={10}/>
					</MemoryRouter>
				</Provider>
			)
			.toJSON();
		expect(tree).toMatchSnapshot();
	});
}),

describe("UpdatePostForm", () => {
	const middlewares = [thunk];
	const mockStore = configureMockStore(middlewares);
	let store;
	store = mockStore(updatePostMock);

	it("renders correctly", () => {
		const tree = renderer
			.create(
				<Provider store={store}>
					<MemoryRouter>
						<UpdatePost min={10}/>
					</MemoryRouter>
				</Provider>
			)
			.toJSON();
		expect(tree).toMatchSnapshot();
	});

	it("basic structure", () => {
		render(
			<Provider store={store}>
				<MemoryRouter>
					<UpdatePost />
				</MemoryRouter>
			</Provider>
		);
		const editTitle = screen.getByTestId('UpdatePostFormTitleLabel');
		const editDescription = screen.getByTestId('UpdatePostFormDescriptionLabel');
		const editUrl = screen.getByTestId('UpdatePostFormUrlLabel');
		const editPrice = screen.getByTestId('UpdatePostFormPriceLabel');
		const editCategory = screen.getByTestId('UpdatePostFormCategoryLabel');
		const editImages = screen.getByTestId('UpdatePostFormImagesLabel');
		const editCode = screen.getByTestId('UpdatePostFormCodeLabel');
	});

	it("edit Title", async () => {
		render(
			<Provider store={store}>
				<MemoryRouter>
					<UpdatePost min={10}/>
				</MemoryRouter>
			</Provider>
		);

		const inputTitle = screen.getByTestId('UpdatePostFormTitleInput');

		fireEvent.change(inputTitle, {
			target: { value: "Rolex" }
		});

		const saveButton = screen.getByTestId('UpdatePostFormSaveButton');

		fireEvent.click(saveButton);

	});

	it("edit Code", async () => {
		render(
			<Provider store={store}>
				<MemoryRouter>
					<UpdatePost min={10}/>
				</MemoryRouter>
			</Provider>
		);

		const inputCode = screen.getByTestId('UpdatePostFormCodeInput');

		fireEvent.change(inputCode, {
			target: { value: "REBAJA" }
		});

		const saveButton = screen.getByTestId('UpdatePostFormSaveButton');

		fireEvent.click(saveButton);

	});

	it("edit Description", async () => {
		render(
			<Provider store={store}>
				<MemoryRouter>
					<UpdatePost min={10}/>
				</MemoryRouter>
			</Provider>
		);

		const inputDescription = screen.getByTestId('UpdatePostFormDescriptionInput');

		fireEvent.change(inputDescription, {
			target: { value: "En buen estado" }
		});

		const saveButton = screen.getByTestId('UpdatePostFormSaveButton');

		fireEvent.click(saveButton);

	});

	it("edit Url", async () => {
		render(
			<Provider store={store}>
				<MemoryRouter>
					<UpdatePost min={10}/>
				</MemoryRouter>
			</Provider>
		);

		const inputUrl = screen.getByTestId('UpdatePostFormUrlInput');

		fireEvent.change(inputUrl, {
			target: { value: "https://wwww.rolex.com" }
		});

		const saveButton = screen.getByTestId('UpdatePostFormSaveButton');

		fireEvent.click(saveButton);

	});

	it("edit Price", async () => {
		render(
			<Provider store={store}>
				<MemoryRouter>
					<UpdatePost min={10}/>
				</MemoryRouter>
			</Provider>
		);

		const inputPrice = screen.getByTestId('UpdatePostFormPriceInput');

		fireEvent.change(inputPrice, {
			target: { value: "15" }
		});

		const saveButton = screen.getByTestId('UpdatePostFormSaveButton');

		fireEvent.click(saveButton);

	});

	it("edit Category", async () => {
		render(
			<Provider store={store}>
				<MemoryRouter>
					<UpdatePost min={10}/>
				</MemoryRouter>
			</Provider>
		);

		const inputCategory = screen.getByTestId('UpdatePostFormCategoryInput');

		fireEvent.change(inputCategory, {
			target: { value: "meals" }
		});

		const saveButton = screen.getByTestId('UpdatePostFormSaveButton');

		fireEvent.click(saveButton);

	});

	it("edit Images", async () => {
		render(
			<Provider store={store}>
				<MemoryRouter>
					<UpdatePost min={10}/>
				</MemoryRouter>
			</Provider>
		);

		const inputImages = screen.getByTestId('UpdatePostFormImagesInput');

		fireEvent.change(inputImages, {
			target: { value: "" }
		});

		const saveButton = screen.getByTestId('UpdatePostFormSaveButton');

		fireEvent.click(saveButton);

	});
});