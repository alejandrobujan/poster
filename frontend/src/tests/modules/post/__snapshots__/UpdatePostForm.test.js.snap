// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpdatePostForm renders correctly 1`] = `
<div>
  <div
    className="container"
  >
    <div
      className="card bg-light border-dark m-4"
    >
      <h5
        className="card-header text-center"
      >
        Update post
      </h5>
      <div
        className="card-body"
      >
        <form
          className="needs-validation"
          noValidate={true}
          onSubmit={[Function]}
        >
          <div
            className="form-group row"
          >
            <label
              className="col-md-3 col-form-label"
              data-testid="UpdatePostFormTitleLabel"
              htmlFor="title"
            >
              Title (*)
            </label>
            <div
              className="col-md-9"
            >
              <input
                autoFocus={true}
                className="form-control"
                data-testid="UpdatePostFormTitleInput"
                id="title"
                maxLength={16}
                minLength={1}
                onChange={[Function]}
                required={true}
                type="text"
                value="Coupon"
              />
              <div
                className="invalid-feedback"
              >
                The title size must be between 1 and 60
              </div>
            </div>
          </div>
          <div
            className="form-group row"
          >
            <label
              className="col-md-3 col-form-label"
              data-testid="UpdatePostFormDescriptionLabel"
              htmlFor="description"
            >
              Description (*)
            </label>
            <div
              className="col-md-9"
            >
              <textarea
                className="form-control"
                data-testid="UpdatePostFormDescriptionInput"
                id="description"
                maxLength={256}
                minLength={1}
                onChange={[Function]}
                required={true}
                value="Buena oferta"
              />
              <div
                className="invalid-feedback"
              >
                The description size must be between 1 and 256
              </div>
            </div>
          </div>
          <div
            className="form-group row"
          >
            <label
              className="col-md-3 col-form-label"
              data-testid="UpdatePostFormUrlLabel"
              htmlFor="url"
            >
              Url
            </label>
            <div
              className="col-md-9"
            >
              <div
                className="input-group"
              >
                <span
                  className="input-group-text"
                >
                  http(s)://
                </span>
                <input
                  className="form-control"
                  data-testid="UpdatePostFormUrlInput"
                  id="url"
                  maxLength={2048}
                  minLength={0}
                  onChange={[Function]}
                  type="text"
                  value="https://cupon"
                />
              </div>
              <div
                className="invalid-feedback"
              >
                The url size must be between 0 and 2048
              </div>
            </div>
          </div>
          <div
            className="form-group row"
          >
            <label
              className="col-md-3 col-form-label"
              data-testid="UpdatePostFormPriceLabel"
              htmlFor="price"
            >
              Price (*)
            </label>
            <div
              className="col-md-3"
            >
              <div
                className="input-group"
              >
                <input
                  className="form-control"
                  data-testid="UpdatePostFormPriceInput"
                  id="price"
                  max="999999.99"
                  min="0.00"
                  onChange={[Function]}
                  required={true}
                  step="0.01"
                  type="number"
                  value={10}
                />
                <span
                  className="input-group-text"
                >
                  â‚¬
                </span>
              </div>
              <div
                className="invalid-feedback"
              >
                Price must be between 0 and 999999.99
              </div>
            </div>
            <label
              className="col-md-3 col-form-label"
              data-testid="UpdatePostFormCategoryLabel"
              htmlFor="categoryId"
            >
              Category
            </label>
            <div
              className="col-md-3"
            >
              <select
                className="form-control"
                data-testid="UpdatePostFormCategoryInput"
                id="categoryId"
                onChange={[Function]}
                value={1}
              >
                <option
                  style={
                    Object {
                      "color": "gray",
                      "fontStyle": "italic",
                    }
                  }
                  value=""
                >
                  Select
                </option>
                <option
                  value={1}
                >
                  Meals
                </option>
                <option
                  value={2}
                >
                  Motor
                </option>
                <option
                  value={3}
                >
                  Home
                </option>
                <option
                  value={4}
                >
                  Toys
                </option>
                <option
                  value={5}
                >
                  Tech
                </option>
                <option
                  value={6}
                >
                  Leisure
                </option>
              </select>
            </div>
          </div>
          <div
            className="form-group row"
          >
            <label
              className="col-md-3 col-form-label"
              data-testid="UpdatePostFormImagesLabel"
              htmlFor="images"
            >
              Images
            </label>
            <div
              className="col-md-9"
            >
              <input
                accept="image/*"
                data-testid="UpdatePostFormImagesInput"
                id="images"
                multiple={true}
                onChange={[Function]}
                type="file"
              />
              <button
                className="btn btn-outline-danger"
                onClick={[Function]}
                style={
                  Object {
                    "display": "none",
                  }
                }
                type="button"
              >
                <svg
                  className="bi bi-trash"
                  fill="currentColor"
                  height="16"
                  viewBox="0 0 16 16"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M5.5 5.5A.5.5 0 0 1 6 6v6a.5.5 0 0 1-1 0V6a.5.5 0 0 1 .5-.5Zm2.5 0a.5.5 0 0 1 .5.5v6a.5.5 0 0 1-1 0V6a.5.5 0 0 1 .5-.5Zm3 .5a.5.5 0 0 0-1 0v6a.5.5 0 0 0 1 0V6Z"
                  />
                  <path
                    d="M14.5 3a1 1 0 0 1-1 1H13v9a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V4h-.5a1 1 0 0 1-1-1V2a1 1 0 0 1 1-1H6a1 1 0 0 1 1-1h2a1 1 0 0 1 1 1h3.5a1 1 0 0 1 1 1v1ZM4.118 4 4 4.059V13a1 1 0 0 0 1 1h6a1 1 0 0 0 1-1V4.059L11.882 4H4.118ZM2.5 3h11V2h-11v1Z"
                  />
                </svg>
              </button>
              <div
                className="invalid-feedback"
              >
                The maximum size allowed is 1MB.
              </div>
            </div>
          </div>
          <div>
            <p>
              New uploaded images:
            </p>
            <p
              className="noImages"
            >
              No new uploaded images.
            </p>
          </div>
          <div>
            <p>
              Previously uploaded images:
            </p>
            <p
              className="noImages"
            >
              No previously uploaded images.
            </p>
          </div>
          <div
            className="form-group row"
          >
            <label
              className="col-md-3 col-form-label"
              data-testid="UpdatePostFormCodeLabel"
              htmlFor="code"
            >
              Code (*)
            </label>
            <div
              className="col-md-9"
            >
              <input
                className="form-control"
                data-testid="UpdatePostFormCodeInput"
                id="code"
                onChange={[Function]}
                required={true}
                type="text"
                value="12345"
              />
              <div
                className="invalid-feedback"
              >
                The code is mandatory
              </div>
            </div>
          </div>
          <div
            className="form-group row"
          >
            <label
              className="col-md-3 col-form-label"
              data-testid="UpdatePostFormExpirationDateLabel"
              htmlFor="code"
            >
              Expiration date (*)
            </label>
            <div
              className="col-md-9"
            >
              <input
                className="form-control"
                data-testid="UpdatePostFormExpirationDateInput"
                id="code"
                min="1970-01-01T01:00"
                onChange={[Function]}
                required={true}
                type="datetime-local"
                value="2022-12-07T11:46"
              />
              <div
                className="invalid-feedback"
              >
                The expiration date is mandatory
              </div>
            </div>
          </div>
          <div>
            <p>
              (*) means mandatory field
            </p>
          </div>
          <div
            className="text-center"
          >
            <button
              className="btn btn-primary"
              data-testid="UpdatePostFormSaveButton"
              type="submit"
            >
              Save changes
            </button>
          </div>
        </form>
        <button
          className="btn btn-primary"
          onClick={[Function]}
        >
          <div
            className="d-flex align-items-center"
          >
            <svg
              className="bi bi-arrow-left"
              fill="currentColor"
              height="16"
              viewBox="0 0 16 16"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M15 8a.5.5 0 0 0-.5-.5H2.707l3.147-3.146a.5.5 0 1 0-.708-.708l-4 4a.5.5 0 0 0 0 .708l4 4a.5.5 0 0 0 .708-.708L2.707 8.5H14.5A.5.5 0 0 0 15 8z"
                fillRule="evenodd"
              />
            </svg>
            <span
              className="ml-2"
            >
              Back
            </span>
          </div>
        </button>
      </div>
    </div>
  </div>
</div>
`;
